Branch-SDK/Branch-SDK/BNCServerInterface.m

```diff
@@ -11,109 +11,130 @@
 #import "BNCConfig.h"
 #import "BNCLinkData.h"
 #import "BNCEncodingUtils.h"
+#import "BNCError.h"
 
 @implementation BNCServerInterface
 
-// make a generalized get request
-- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag {
-    [self getRequestAsync:params url:url andTag:requestTag retryNumber:0 log:YES];
-}
+#pragma mark - GET methods
 
-// this is actually a synchronous call; it should NOT be called from the main queue
-- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag log:(BOOL)log {
-    [self getRequestAsync:params url:url andTag:requestTag retryNumber:0 log:log];
+- (void)getRequest:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag callback:(BNCServerCallback)callback {
+    [self getRequest:params url:url andTag:requestTag retryNumber:0 log:YES callback:callback];
 }
 
-- (void)getRequestAsync:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag retryNumber:(NSInteger)retryNumber log:(BOOL)log {
-    BNCServerResponse *serverResponse = [self genericSyncHTTPRequest:[self prepareGetRequest:params url:url retryNumber:retryNumber log:log] withTag:requestTag andLinkData:nil];
+- (void)getRequest:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag log:(BOOL)log callback:(BNCServerCallback)callback {
+    [self getRequest:params url:url andTag:requestTag retryNumber:0 log:log callback:callback];
+}
 
-    NSInteger status = [serverResponse.statusCode integerValue];
-    BOOL isRetryableStatusCode = status >= 500;
-    
-    // Retry the request if appropriate
-    if (retryNumber < [BNCPreferenceHelper getRetryCount] && isRetryableStatusCode) {
-        [NSThread sleepForTimeInterval:[BNCPreferenceHelper getRetryInterval]];
+- (void)getRequest:(NSDictionary *)params url:(NSString *)url andTag:(NSString *)requestTag retryNumber:(NSInteger)retryNumber log:(BOOL)log callback:(BNCServerCallback)callback {
+    NSURLRequest *request = [self prepareGetRequest:params url:url retryNumber:retryNumber log:log];
 
-        if (log) {
-            [BNCPreferenceHelper log:FILE_NAME line:LINE_NUM message:@"Replaying request with tag %@", requestTag];
-        }
-
-        [self getRequestAsync:params url:url andTag:requestTag retryNumber:++retryNumber log:log];
-    }
-    // Otherwise, let the delegate handle it
-    else if (self.delegate) {
-        [self.delegate serverCallback:serverResponse];
-    }
+    [self genericHTTPRequest:request withTag:requestTag andLinkData:nil retryNumber:retryNumber log:log callback:callback retryHandler:^NSURLRequest *(NSInteger lastRetryNumber) {
+        return [self prepareGetRequest:params url:url retryNumber:++lastRetryNumber log:log];
```

Comment:
Yeah, I know. I'm not a fan of it either, but since the URL Request includes the actual request body / query string, it has to be separate, unless you've got any better thoughts for that.